miner/algo_greedy.go

```diff
@@ -20,23 +20,58 @@ type greedyBuilder struct {
 	chainData        chainData
 	builderKey       *ecdsa.PrivateKey
 	interrupt        *int32
+	buildBlockFunc   BuildBlockFunc
 	algoConf         algorithmConfig
 }
 
 func newGreedyBuilder(
 	chain *core.BlockChain, chainConfig *params.ChainConfig, algoConf *algorithmConfig,
 	blacklist map[common.Address]struct{}, env *environment, key *ecdsa.PrivateKey, interrupt *int32,
-) *greedyBuilder {
+) (*greedyBuilder, error) {
 	if algoConf == nil {
-		algoConf = &defaultAlgorithmConfig
+		return nil, errNoAlgorithmConfig
 	}
-	return &greedyBuilder{
+
+	builder := &greedyBuilder{
 		inputEnvironment: env,
-		chainData:        chainData{chainConfig, chain, blacklist},
+		chainData:        chainData{chainConfig: chainConfig, chain: chain, blacklist: blacklist},
 		builderKey:       key,
 		interrupt:        interrupt,
 		algoConf:         *algoConf,
 	}
+	// Initialize block builder function
+	var buildBlockFunc BuildBlockFunc
+	if algoConf.EnableMultiTxSnap {
+		buildBlockFunc = func(simBundles []types.SimulatedBundle, simSBundles []*types.SimSBundle, transactions map[common.Address]types.Transactions) (*environment, []types.SimulatedBundle, []types.UsedSBundle) {
```

Comment:
this was a stepping stone towards that, i didn't want to add more changes to this PR but I was planning on that approach. I can include it.

Reply:
updated to de-dupe logic 

It's not great and open to other suggestions but might refactor more in separate PR since this one is quite large

I'm open to just creating a separate efficient revert builder but that can be less flexible because we might want to test correctness when ordering logic is different for builders

Reply:
ended up just creating separate builders that have multi snap, removed it from configuration

we'd need to refactor this code to handle more dynamic configurations and i don't want to do that in this PR